src/jdbc/cubrid/jdbc/jci/UTimedDataInputStream.java

```diff
@@ -49,101 +49,118 @@
+        long begin = System.currentTimeMillis();
+
+        while (true) {
+          try {
+            stream.readFully(b);
+            return;
+          } catch (SocketTimeoutException e) {
+            if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
+              String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
+              throw new SocketTimeoutException(msg);
+            }
+            BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
+          }
+        }
     }
 
-    public int readByte(byte[] b, int timeout) throws IOException {
-	long begin = System.currentTimeMillis();
-
-	while (true) {
-	    try {
-		return stream.read(b);
-	    } catch (SocketTimeoutException e) {
-		if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
-		    String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
-		    throw new SocketTimeoutException(msg);
-		}
-		BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
-	    }
-	}
+    public int readByte(byte[] b, int timeout) throws IOException, UJciException {
+        long begin = System.currentTimeMillis();
+
+        while (true) {
+          try {
+            return stream.read(b);
+          } catch (SocketTimeoutException e) {
+            if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
+              String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
+              throw new SocketTimeoutException(msg);
+            }
+            BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
+          }
+        }
     }
 
-    public int readByte(byte[] b) throws IOException {
-	return readByte(b, timeout);
+    public int readByte(byte[] b) throws IOException, UJciException {
+        return readByte(b, timeout);
     }
 
-    public int read(byte[] b, int off, int len, int timeout) throws IOException {
-	long begin = System.currentTimeMillis();
-
-	while (true) {
-	    try {
-		return stream.read(b, off, len);
-	    } catch (SocketTimeoutException e) {
-		if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
-		    String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
-		    throw new SocketTimeoutException(msg);
-		}
-		BrokerHandler.pingBroker(ip, port, PING_TIMEOUT);
-	    }
-	}
+    public int read(byte[] b, int off, int len, int timeout) throws IOException, UJciException {
+        long begin = System.currentTimeMillis();
+
+        while (true) {
+          try {
+            return stream.read(b, off, len);
+          } catch (SocketTimeoutException e) {
+            if (timeout > 0 && timeout - (System.currentTimeMillis() - begin) <= 0) {
+              String msg = UErrorCode.codeToMessage(UErrorCode.ER_TIMEOUT);
+              throw new SocketTimeoutException(msg);
+            }
+            if (BrokerHandler.statusBroker(ip, port, pid, session, PING_TIMEOUT) != 1) {
+              String msg = UErrorCode.codeToMessage(UErrorCode.ER_COMMUNICATION);
```

Comment:
how about using `throw new UJciException(UErrorCode.ER_COMMUNICATION);` instead of java built-in IOException to distinguish that socket is closed or that CAS stays in the wrong status because of abnormal network.

Reply:
I revised the code which you commented.